src/modules/networkmanager.rs

```diff
@@ -57,52 +211,115 @@ impl Module<GtkBox> for NetworkManagerModule {
 
-        glib::spawn_future_local({
+        let container_clone = container.clone();
+        context.subscribe().recv_glib_async((), move |(), update| {
+            let container = container.clone();
             let image_provider = image_provider.clone();
-            let icon = icon.clone();
-
+            let this = self.clone();
             async move {
-                image_provider
-                    .load_into_picture_silent(INITIAL_ICON_NAME, self.icon_size, false, &icon)
-                    .await;
-            }
-        });
+                match update {
+                    NetworkManagerUpdate::Devices(devices) => {
+                        tracing::debug!("NetworkManager devices updated");
+                        tracing::trace!("NetworkManager devices updated: {devices:#?}");
 
-        context.subscribe().recv_glib_async((), move |(), state| {
-            let image_provider = image_provider.clone();
-            let icon = icon.clone();
-
-            let icon_name = match state {
-                ClientState::WiredConnected => "network-wired-symbolic",
-                ClientState::WifiConnected => "network-wireless-symbolic",
-                ClientState::CellularConnected => "network-cellular-symbolic",
-                ClientState::VpnConnected => "network-vpn-symbolic",
-                ClientState::WifiDisconnected => "network-wireless-acquiring-symbolic",
-                ClientState::Offline => "network-wireless-disabled-symbolic",
-                ClientState::Unknown => "dialog-question-symbolic",
-            };
+                        // resize the container's children to match the number of devices
+                        if container.children().count() > devices.len() {
+                            for child in container.children().skip(devices.len()) {
+                                container.remove(&child);
+                            }
+                        } else {
+                            while container.children().count() < devices.len() {
+                                let icon = Picture::builder()
+                                    .content_fit(ContentFit::ScaleDown)
+                                    .css_classes(["icon"])
+                                    .build();
+                                container.append(&icon);
+                            }
+                        }
 
-            async move {
-                image_provider
-                    .load_into_picture_silent(icon_name, self.icon_size, false, &icon)
-                    .await;
+                        // update each icon to match the device state
+                        for (device, widget) in devices.iter().zip(container.children()) {
+                            this.update_icon(
+                                &image_provider,
+                                device,
+                                widget.downcast_ref::<Picture>().unwrap(),
```

Comment:
CI should be flagging `unwrap`, I'm not sure why it isn't

```suggestion
                                widget.downcast_ref::<Picture>().expect("widget should be picture"),
```

Same below too pls